subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -150,13 +150,33 @@ protected void compile(InputChanges inputChanges) {
         checkGroovyClasspathIsNonEmpty();
         warnIfCompileAvoidanceEnabled();
 
-        if (inputChanges != null && getOptions().isIncremental()) {
-            doIncrementalCompile(inputChanges);
+        GroovyJavaJointCompileSpec spec = createSpec();
+
+        if (inputChanges != null && enableIncrementalCompilation(spec)) {
+            doIncrementalCompile(spec, inputChanges);
         } else {
-            doCompile(inputChanges, null);
+            doCompile(spec, inputChanges, null);
+        }
+    }
+
+    private void doIncrementalCompile(GroovyJavaJointCompileSpec spec, InputChanges inputChanges) {
+        Multimap<String, String> oldMappings = readSourceClassesMappingFile(getSourceClassesMappingFile());
+        getSourceClassesMappingFile().delete();
+
+        WorkResult result = doCompile(spec, inputChanges, oldMappings);
+
+        if (!isFullRecompilation(result)) {
```

Comment:
If it was a full recompilation, we still need to write all the new mappings to the file.

Reply:
I think if it's a full recompilation, the mappings are already written in the compilation process by the Groovy compiler plugin in `ApiGroovyCompiler`, right?

Reply:
Yeah, you are right. So we don't need to merge the two mappings files. Can we add a comment about this somewhere? It is not obvious from the code here.

Reply:
Done.